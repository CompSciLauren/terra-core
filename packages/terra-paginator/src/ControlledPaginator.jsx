import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import ResponsiveElement from 'terra-responsive-element';
import IconPrevious from 'terra-icon/lib/icon/IconPrevious';
import IconNext from 'terra-icon/lib/icon/IconNext';
import KeyCode from 'keycode-js';
import styles from './Paginator.module.scss';
import Button from '../../terra-button';
import { calculatePages, pageSet } from './_paginationUtils';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Function to be executed when a navigation element is selected.
   */
  onPageChange: PropTypes.func.isRequired,
  /**
   * The active/selected page. Used to set the default selected page or maintain selection across renders.
   * Required when using totalCount and itemCountPerPage.
   */
  selectedPage: PropTypes.number.isRequired,
  /**
   * Total number of all items being paginated.
   * Required when using itemCountPerPage and selectedPage.
   */
  totalCount: PropTypes.number.isRequired,
  /**
   * Total number of items per page.
   * Required when using selectedPage and totalCount.
   */
  itemCountPerPage: PropTypes.number.isRequired,
};

const contextTypes = {
  /* eslint-disable consistent-return */
  intl: (context) => {
    if (context.intl === undefined) {
      return new Error('Component is internationalized, and must be wrapped in terra-base');
    }
  },
};

class Paginator extends React.Component {
  constructor(props) {
    super(props);

    this.handlePageChange = this.handlePageChange.bind(this);
    this.handleOnKeyDown = this.handleOnKeyDown.bind(this);
    this.hasNavContext = this.hasNavContext.bind(this);
    this.buildPageButtons = this.buildPageButtons.bind(this);
    this.reducedPaginator = this.reducedPaginator.bind(this);
  }

  handlePageChange(index) {
    return (event) => {
      event.preventDefault();

      if (Number.isNaN(Number(index))) {
        this.props.onPageChange(event.currentTarget.attributes['aria-label'].value);

        return false;
      }

      this.props.onPageChange(index);

      return false;
    };
  }

  handleOnKeyDown(index) {
    return (event) => {
      if (event.nativeEvent.keyCode === KeyCode.KEY_RETURN || event.nativeEvent.keyCode === KeyCode.KEY_SPACE) {
        event.preventDefault();

        if (Number.isNaN(Number(index))) {
          this.props.onPageChange(event.currentTarget.attributes['aria-label'].value);

          return false;
        }

        this.props.onPageChange(index);
      }

      return false;
    };
  }

  buildPageButtons(totalPages, onClick) {
    const { totalCount, itemCountPerPage, selectedPage } = this.props;
    const pageSequence = pageSet(selectedPage, calculatePages(totalCount, itemCountPerPage));
    const pageButtons = [];

    pageSequence.forEach((val) => {
      const paginationLinkClassNames = cx([
        'nav-link',
        val === selectedPage ? 'is-selected' : null,
      ]);

      if (val > totalPages) {
        return;
      }
      pageButtons.push((
        <Button
          aria-label={`Page ${val}`}
          aria-current={val === selectedPage && 'page'}
          className={paginationLinkClassNames}
          tabIndex={val === selectedPage ? null : '0'}
          key={`pageButton_${val}`}
          onClick={onClick(val)}
          onKeyDown={this.handleOnKeyDown(val)}
          text={String(val)}
        />
      ));
    });

    return pageButtons;
  }

  hasNavContext() {
    return this.props.totalCount && this.props.itemCountPerPage;
  }

  defaultPaginator() {
    const totalPages = calculatePages(this.props.totalCount, this.props.itemCountPerPage);
    const { selectedPage } = this.props;
    const previousPageIndex = selectedPage === 1 ? 1 : selectedPage - 1;
    const nextPageIndex = selectedPage === totalPages ? totalPages : selectedPage + 1;
    const areLeftControlsDisabled = selectedPage === 1;
    const areRightControlsDisabled = selectedPage === totalPages;

    const fullView = (
      <div className={cx(['paginator', !this.hasNavContext() && 'pageless'])}>
        {
          this.hasNavContext() && (
            <Button
              className={cx(['nav-link', areLeftControlsDisabled && 'is-disabled'])}
              isDisabled={areLeftControlsDisabled}
              aria-disabled={areLeftControlsDisabled}
              aria-label="first"
              tabIndex={areLeftControlsDisabled ? null : '0'}
              onClick={this.handlePageChange(1)}
              onKeyDown={this.handleOnKeyDown(1)}
              text={this.context.intl.formatMessage({ id: 'Terra.paginator.first' })}
            />
          )
        }
        <Button
          className={cx(['nav-link', 'previous', areLeftControlsDisabled && 'is-disabled'])}
          icon={<IconPrevious />}
          isDisabled={areLeftControlsDisabled}
          aria-disabled={areLeftControlsDisabled}
          aria-label="previous"
          tabIndex={areLeftControlsDisabled ? null : '0'}
          onClick={this.handlePageChange(previousPageIndex)}
          onKeyDown={this.handleOnKeyDown(previousPageIndex)}
          text={this.context.intl.formatMessage({ id: 'Terra.paginator.previous' })}
        />
        {this.hasNavContext() && this.buildPageButtons(totalPages, this.handlePageChange)}
        <Button
          className={cx(['nav-link', 'next', areRightControlsDisabled && 'is-disabled'])}
          icon={<IconNext />}
          isReversed
          isDisabled={areRightControlsDisabled}
          aria-disabled={areRightControlsDisabled}
          aria-label="next"
          tabIndex={areRightControlsDisabled ? null : '0'}
          onClick={this.handlePageChange(nextPageIndex)}
          onKeyDown={this.handleOnKeyDown(nextPageIndex)}
          text={this.context.intl.formatMessage({ id: 'Terra.paginator.next' })}
        />
        {
          this.hasNavContext() && (
            <Button
              className={cx(['nav-link', areRightControlsDisabled && 'is-disabled'])}
              isDisabled={areRightControlsDisabled}
              aria-disabled={areRightControlsDisabled}
              aria-label="last"
              tabIndex={areRightControlsDisabled ? null : '0'}
              onClick={this.handlePageChange(totalPages)}
              onKeyDown={this.handleOnKeyDown(totalPages)}
              text={this.context.intl.formatMessage({ id: 'Terra.paginator.last' })}
            />
          )
        }
      </div>
    );

    return fullView;
  }

  reducedPaginator() {
    const totalPages = calculatePages(this.props.totalCount, this.props.itemCountPerPage);
    const { selectedPage } = this.props;
    const previousPageIndex = selectedPage === 1 ? 1 : selectedPage - 1;
    const nextPageIndex = selectedPage === totalPages ? totalPages : selectedPage + 1;
    const areLeftControlsDisabled = selectedPage === 1;
    const areRightControlsDisabled = selectedPage === totalPages;

    const reducedView = (
      <div className={cx(['paginator', !this.hasNavContext() && 'pageless'])} role="navigation" aria-label="pagination">
        {
          this.hasNavContext() && (
            <Button
              className={cx(['nav-link', areLeftControlsDisabled && 'is-disabled'])}
              isDisabled={areLeftControlsDisabled}
              aria-disabled={areLeftControlsDisabled}
              aria-label="first"
              tabIndex={areLeftControlsDisabled ? null : '0'}
              onClick={this.handlePageChange(1)}
              onKeyDown={this.handleOnKeyDown(1)}
              text={this.context.intl.formatMessage({ id: 'Terra.paginator.first' })}
            />
          )
        }
        <Button
          className={cx(['nav-link', 'previous', areLeftControlsDisabled && 'is-disabled'])}
          icon={<IconPrevious />}
          isIconOnly
          isDisabled={areLeftControlsDisabled}
          aria-disabled={areLeftControlsDisabled}
          aria-label="previous"
          tabIndex={areLeftControlsDisabled ? null : '0'}
          onClick={this.handlePageChange(previousPageIndex)}
          onKeyDown={this.handleOnKeyDown(previousPageIndex)}
          text={this.context.intl.formatMessage({ id: 'Terra.paginator.previous' })}
        />
        {this.hasNavContext() && this.context.intl.formatMessage({ id: 'Terra.paginator.pageIndex' }, { selectedPage })}
        <Button
          className={cx(['nav-link', 'next', areRightControlsDisabled && 'is-disabled'])}
          icon={<IconNext />}
          isIconOnly
          isDisabled={areRightControlsDisabled}
          aria-disabled={areRightControlsDisabled}
          aria-label="next"
          tabIndex={areRightControlsDisabled ? null : '0'}
          onClick={this.handlePageChange(nextPageIndex)}
          onKeyDown={this.handleOnKeyDown(nextPageIndex)}
          text={this.context.intl.formatMessage({ id: 'Terra.paginator.next' })}
        />
        {
          this.hasNavContext() && (
            <Button
              className={cx(['nav-link', areRightControlsDisabled && 'is-disabled'])}
              isDisabled={areRightControlsDisabled}
              aria-disabled={areRightControlsDisabled}
              aria-label="last"
              tabIndex={areRightControlsDisabled ? null : '0'}
              onClick={this.handlePageChange(totalPages)}
              onKeyDown={this.handleOnKeyDown(totalPages)}
              text={this.context.intl.formatMessage({ id: 'Terra.paginator.last' })}
            />
          )
        }
      </div>
    );

    return reducedView;
  }

  render() {
    return <ResponsiveElement defaultElement={this.reducedPaginator()} small={this.defaultPaginator()} />;
  }
}

Paginator.propTypes = propTypes;
Paginator.contextTypes = contextTypes;

export default Paginator;
