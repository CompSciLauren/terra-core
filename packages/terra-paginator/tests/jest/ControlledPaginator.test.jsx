import React from 'react';
/* eslint-disable-next-line import/no-extraneous-dependencies */
import { mountWithIntl } from 'terra-enzyme-intl';

import ControlledPaginator from '../../src/ControlledPaginator';

describe('Paginator', () => {
  it('should render a controlled paginator', () => {
    const wrapper = mountWithIntl(<ControlledPaginator onPageChange={e => typeof e} selectedPage={1} totalCount={2234} itemCountPerPage={20} />);
    expect(wrapper).toMatchSnapshot();
  });
});
