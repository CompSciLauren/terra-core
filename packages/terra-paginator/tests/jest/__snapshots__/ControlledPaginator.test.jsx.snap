// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator should render a Controlled Paginator 1`] = `
<ResponsiveElement
  responsiveTo="parent"
  small={
    <div
      className="paginator"
    >
      <PaginatorButton
        aria-disabled={true}
        aria-label="Terra.paginator.first"
        className="nav-link left-controls is-disabled"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={null}
        type="button"
      >
        Terra.paginator.first
      </PaginatorButton>
      <PaginatorButton
        aria-disabled={true}
        aria-label="Terra.paginator.previous"
        className="nav-link left-controls previous is-disabled"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={null}
        type="button"
      >
        <span
          className="icon"
        />
        Terra.paginator.previous
      </PaginatorButton>
      <PaginatorButton
        aria-current="page"
        aria-label="Terra.paginator.pageIndex"
        className="nav-link is-selected"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={null}
        type="button"
      >
        1
      </PaginatorButton>
      <PaginatorButton
        aria-current={false}
        aria-label="Terra.paginator.pageIndex"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        2
      </PaginatorButton>
      <PaginatorButton
        aria-current={false}
        aria-label="Terra.paginator.pageIndex"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        3
      </PaginatorButton>
      <PaginatorButton
        aria-current={false}
        aria-label="Terra.paginator.pageIndex"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        4
      </PaginatorButton>
      <PaginatorButton
        aria-current={false}
        aria-label="Terra.paginator.pageIndex"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        5
      </PaginatorButton>
      <PaginatorButton
        aria-current={false}
        aria-label="Terra.paginator.pageIndex"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        6
      </PaginatorButton>
      <PaginatorButton
        aria-current={false}
        aria-label="Terra.paginator.pageIndex"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        7
      </PaginatorButton>
      <PaginatorButton
        aria-current={false}
        aria-label="Terra.paginator.pageIndex"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        8
      </PaginatorButton>
      <PaginatorButton
        aria-current={false}
        aria-label="Terra.paginator.pageIndex"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        9
      </PaginatorButton>
      <PaginatorButton
        aria-current={false}
        aria-label="Terra.paginator.pageIndex"
        className="nav-link"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        10
      </PaginatorButton>
      <PaginatorButton
        aria-disabled={false}
        aria-label="Terra.paginator.next"
        className="nav-link right-controls next"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        Terra.paginator.next
        <span
          className="icon"
        />
      </PaginatorButton>
      <PaginatorButton
        aria-disabled={false}
        aria-label="Terra.paginator.last"
        className="nav-link right-controls"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        Terra.paginator.last
      </PaginatorButton>
    </div>
  }
  tiny={
    <div
      aria-label="pagination"
      className="paginator"
      role="navigation"
    >
      <PaginatorButton
        aria-disabled={true}
        aria-label="Terra.paginator.first"
        className="nav-link left-controls is-disabled"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={null}
        type="button"
      >
        Terra.paginator.first
      </PaginatorButton>
      <PaginatorButton
        aria-disabled={true}
        aria-label="Terra.paginator.previous"
        className="nav-link left-controls previous is-disabled"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={null}
        type="button"
      >
        <span
          className="icon"
        />
        Terra.paginator.previous
      </PaginatorButton>
      Terra.paginator.pageIndex
      <PaginatorButton
        aria-disabled={false}
        aria-label="Terra.paginator.next"
        className="nav-link right-controls next icon-only"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        <VisuallyHiddenText
          text="Terra.paginator.next"
        />
        <span
          className="icon"
        />
      </PaginatorButton>
      <PaginatorButton
        aria-disabled={false}
        aria-label="Terra.paginator.last"
        className="nav-link right-controls"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex="0"
        type="button"
      >
        Terra.paginator.last
      </PaginatorButton>
    </div>
  }
/>
`;
